### ISSUE TITLE
New op request: ttnn.scatter_add

### ISSUE THREAD

**Issue by sraizada-tt** [tenstorrent/tt-metal#32523]
Parent issue: https://github.com/tenstorrent/tt-metal/issues/32212
need a ttnn version of torch.scatter_add https://docs.pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_

with input size [32, vocab_size (multiple of 32)] on dim 1

**Comment by ekuslysTT** [comment]
@cmaryanTT please assign someone from your team

**Comment by cmaryanTT** [comment]
@sraizada-tt - we can mock this up with existing OPs for now. Something like:
z = ttnn.zeros()
addend = ttnn.scatter(z, dim, inx, src)
out = ttnn.add(x, addend)

@jbbieniekTT - FYI

**Comment by sraizada-tt** [comment]
I don't think this will work 

scatter overwrites values rather than adding them. If multiple indices point to the same location (same token), only the last value is kept (no accumulation)

This can work for decode because there we have a single token each iteration 
Won't work for prefill since the prompt can have multiple counts of the same token 

For now, if you want we can make the prefill histogram fallback to host (since it only needs to be done once) and do this for decode (needs to run each iteration)

What do you think?

**Comment by cmaryanTT** [comment]
Ah, I see, you're expecting non-unique indices

**Comment by cmaryanTT** [comment]
@sraizada-tt - can you please post the vocab size here, and really the full call details of your test case (shapes, params)

@jbbieniekTT has most of the functionality to do an add reduction in scatter that could solve this

**Comment by sraizada-tt** [comment]
B = 32, V = 128*1024

counts: int32 tensor, shape [B, V],  (TILE_LAYOUT). 
idx: int32 tensor, shape [B, S] (S = number of new tokens this step = 1 for decode, >1 for prefill). 
updates: int32  scalar. 1.0
dim: fixed to 1 (vocab axis); we don’t need other axes.


equivalent torch code:
counts[:, :] = counts.scatter_add(1, idx, updates)


or 
        new_tokens = ttnn.to_torch(ttnn.get_device_tensors(new_tokens)[0]).to(torch.int64)
        counts_new = torch.zeros((32, 128 * 1024), dtype=torch.int32)
        updates = torch.ones_like(new_tokens, dtype=torch.int32)
        counts_new = counts_new.scatter_add(dim=1, index=new_tokens, src=updates)
        counts_new = ttnn.from_torch(counts_new, dtype=ttnn.int32, layout=ttnn.TILE_LAYOUT, device=self.mesh_device)

**Comment by cmaryanTT** [comment]
@sraizada-tt can we do counts_new in RM instead of TILE? This bypasses an issue with the scatter OP (we don't have an int untilize/tilize, and the scatter OP is natively RM) - otherwise we need an LLK fix

**Comment by sraizada-tt** [comment]
we can untilize counts_new before sending it to the op

looks like pack_untilize supports uint32 inputs: https://github.com/tenstorrent/tt-metal/blob/30fd522e4d8e8bf6fd5d675a375166c6dbfedecc/ttnn/cpp/ttnn/operations/data_movement/untilize/device/untilize_op.cpp#L168

**Comment by sraizada-tt** [comment]
also please add sub_core_grid support to this op while you are bringing it up

**Comment by jbbieniekTT** [comment]
@sraizada-tt there is no constraint against using it in the code, indeed, but it does not always work as expected - take the `[32, 128*1024]` shape as an example:


```
import ttnn
import torch
import random

from tests.ttnn.utils_for_testing import assert_allclose

dev = ttnn.open_device(device_id=0)

data = [i for i in range(32*128*1024)]

a = ttnn.Tensor(data, [32, 128*1024], data_type=ttnn.int32, layout=ttnn.TILE_LAYOUT, device=dev)
b = ttnn.to_layout(a, ttnn.ROW_MAJOR_LAYOUT)

print("comparison")
print(ttnn.sum(ttnn.typecast(a == b, ttnn.float32)))
print(f"32*128*1024 == {32*128*1024}")
```


the output:
```
comparison
ttnn.Tensor(4190208.0000, shape=Shape([]), dtype=DataType::FLOAT32, layout=Layout::TILE)
32*128*1024 == 4194304
```

float provides reliable `ULP == 1.0` precision until 24 RHS bits of corresponding integers have been reached - here, `32*128*1024 == 2^22`.

also, you can uncomment one of the cases in `test_scatter.py`:
`([1, 151936], -1, [1, 151936], ttnn.bfloat16, ttnn.int32, ttnn.Layout.TILE),`

we cannot rely on untilizing for the time being.

**Comment by sraizada-tt** [comment]
^ I ran this untilize for bfloat16 input and it works
so how about we make counts_new bf16?

**Comment by cmaryanTT** [comment]
@sraizada-tt - we looked at this only as int because this is basically a histogram. If you need this to maintain a stable count over a long window of tokens, it needs to be int. We might be able to get away with using float types, but we have to be careful:
- bf16 gives precise ints up to 7 bits (128)
- fp32 gives precise ints up to 23 bits (8M)
- But, we have to be careful what we do with fp32, because any operation going through the FPU will remap to tf32, trimming the mantissa to 10 bits (1024)

All that said, let's not overthink this. @jbbieniekTT - keep going and we can validate Stuti's test exact test cases after we get a basic implementation done.

**Comment by jbbieniekTT** [comment]
hello, im working on `scatter_add`, im implementing the `sub_core_grid` parameter

**Comment by sraizada-tt** [comment]
question: 
while testing out integration with other sub core grid ops I found that reshaping the output token of decode from [batch_size] to [batch_size, 1] is not a no-op 

which will require reshape to support sub_core_grids

is it straightforward to support idx of size [batch_size] instead of [batch_size, 1] in this op? so that we don't have to do the reshape

**Comment by ekuslysTT** [comment]
Notes
- PR expected by EOD

**Comment by jbbieniekTT** [comment]
@sraizada-tt https://github.com/tenstorrent/tt-metal/pull/33075

**Comment by ekuslysTT** [comment]
@sraizada-tt to link separate issue

**Comment by sraizada-tt** [comment]
issue for oom for llama: https://github.com/tenstorrent/tt-metal/issues/33245
